Add SU(N) LSWT to Sunny

Hao-Phys opened the pull request:
This PR adds the SU(N) linear spin-wave theory features to Sunny. The `:SUN` and `: dipole` modes are supported. But the `:large_S` mode is not supported yet. Moreover, the `biquad` interactions are not supported.

kbarros commented on src/Sunny.jl:
```diff
@@ -89,6 +89,16 @@ export DynamicStructureFactor, InstantStructureFactor, StructureFactor, FormFact
+export SpinWaveFields, construct_magnetic_supercell,
+    set_dipoles!, lswt_dispersion_relation, lswt_dynamical_spin_structure_factor!, lswt_unpolarized_INS_spec
+
+# include("WangLandau/BinnedArray.jl")
+# include("WangLandau/WangLandau.jl")
+# export BinnedArray, filter_visited, reset!,
+#     WangLandau, spherical_cap_update, init_bounded!, run!
```
You probably noticed WangLandau/ got deleted -- these lines are no longer necessary, please remove.

Hao-Phys replied:
I have removed these lines.

kbarros replied:
Thanks a lot. On this topic, maybe we should replace the name `polarize_spins!` with something more symmetric with `set_coherent_state!`

kbarros commented on test/test_lswt.jl:
```diff
@@ -0,0 +1,109 @@
+    λ  = 0.1
+    langevin = Langevin(Δt; kT=0, λ)
+
+    dipoles = S * [1.0 1.0 -1.0 -1.0; 1.0 -1.0 -1.0 1.0; 1.0 -1.0 1.0 -1.0] / √3
+    positions = [0.0 0.5 0.5 0.0; 0.0 0.5 0.0 0.5; 0.0 0.0 0.5 0.5]
+    set_dipoles!(sys, dipoles, positions)
```
Are you ok with replacing this as:
```
polarize_spin!(sys, (1, 1, 1), position_to_site(sys, (0, 0, 0)))
polarize_spin!(sys, (1, -1, -1), position_to_site(sys, (1/2, 1/2, 0)))
...
```

Instead of `polarize_spin!` there is also `set_coherent_state!` which functions similarly.

Hao-Phys replied:
`polarize_spin!` looks cleaner than `set_dipoles`. I have updated these lines and removed `set_dipoles`.

kbarros commented on src/SpinWave/SWUtil.jl:
```diff
@@ -0,0 +1,236 @@
+            T̃_mat[:, :, site] = Hermitian(U_mat_N' * sys.interactions_union[site].aniso.matrep * U_mat_N)[1:2, 1:2]
+        end
+    
+    # here I need help from Kipton and David to map back the solution, because in general this is difficult to implement.
+    elseif sys.mode == :large_S
+        error("unsupported")
+        # spin  = (N-1) / 2
+        # s_mat_2 = spin * spin_matrices(2)
```
I'm happy to look at this. I wonder if we should simply 'abort' if certain complicated terms are present here (e.g. scalar biquadratic).

kbarros replied:
Following our conversation, let's remove `:large_S` and instead provide instructions for people to perform their own anti-renormalization?

kbarros commented:
Hi Hao, thanks for the PR. Overall this looks great to me. My suggestions are mostly concerning the public-facing user interface. Maybe we should discuss together with @ddahlbom . The implementation details can always be revised later, so I want to focus for now one the question of what exported functions we should be exposing.

ddahlbom reviewed (commented):
Thanks a lot, Hao. This looks great.

I added a few comments about naming. In general, I think we may want to replace instances of `lswt` with `spin_wave`, which would be more consistent with the existing naming philosophy in Sunny.

I also added some notes regarding how this may be refactored a bit so that it will be well integrated with the existing structure factor code, but I think this can be further developed after merging.

It might be nice to add an example to the FeI2 example before merging, as Kipton mentioned. I'm happy to do this, if you'd like, or I can show you how to modify the example.

kbarros commented on src/SpinWave/SpinWave.jl:
```diff
@@ -0,0 +1,408 @@
+end
+
+
+@inline lorentzian(x :: Float64, η :: Float64) = η / (π * (x^2 + η^2))
```
I agree there's not an obvious place. Shall we keep the one in StructureFactor for now?

ddahlbom replied:
That sounds sensible.

kbarros commented on src/SpinWave/SpinWave.jl:
```diff
@@ -0,0 +1,408 @@
+
+Computes the spin excitation energy dispersion relations given a `SpinWaveField` and `k`. Note that `k` is a 3-vector, the units of kᵢ is 2π/|aᵢ|, where |aᵢ| is the lattice constant of the **chemical** lattice.
+"""
+function lswt_dispersion_relation(sw_fields :: SpinWaveFields, k :: Vector{Float64})
```
I guess if it's a function that will be exposed to end-users, the brevity of the name `lswt_dispersion` has a certain benefit (vs a classical `dispersion` function?). A possible concern is that the name `spin_wave_dispersion` could imply either classical or linearized.

ddahlbom replied:
I do like the brevity. I just wonder whether the full phrase "linear spin wave theory" is something that the user will automatically decode, or whether `lswt` it will just be a sequence of characters that has to be remembered. 

The "linearized" versus "not linearized" distinction could also break down somewhat if we one day ever add higher-order corrections to Hao's code. 

I understand that the classical dynamics can also be used to study spin waves in a sense, but I have the impression that "spin wave theory" does have an understood meaning in the magnetism community, referring to the quantized linearized dynamics and its extensions. What about just using `sw_dispersion` or `swt_dispersion`? 

I guess there's also a question of whether the existing function should be given a new prefix. We could have a "classical/cl" vs "spin_wave/sw" distinction throughout the code.

Just thinking out loud here.

kbarros replied:
Good points. Your comment makes me wonder whether LSWT will become the default, and classical is the thing we prefix with `cl_`. I'm not sure at all, just trying to imagine how most people will be using this eventually. Given that loop corrections might also come, it does seem like the distinction is between classical and QM.

kbarros replied:
After writing that out, I don't like the way `cl_` looks either. For `dispersion` there is no chance of ambiguity. For `intensities` the result will depend of the types given as input. Does this resolve it?

ddahlbom replied:
Yes, I like that (i.e., just having `dispersion` and `intensities`, and having the latter dispatch on type.)

Hao-Phys commented on src/SpinWave/SpinWave.jl:
```diff
@@ -0,0 +1,408 @@
+Sαβ_matrix[:, 7:9] → 2*imag(xy-yx), 2*imag(yz-zy), 2*imag(zx-xz). \n 
+Note that `k` is a 3-vector, the units of kᵢ is 2π/|aᵢ|, where |aᵢ| is the lattice constant of the **chemical** lattice.
+"""
+function lswt_dynamical_spin_structure_factor!(sw_fields :: SpinWaveFields, k :: Vector{Float64}, disp :: Vector{Float64}, Sαβ_matrix :: Matrix{Float64})
```
1.  We can make `Sαβ_matrix` allocated inside the function.  We can make it consistent with the structure of the classical dynamics.
2.  The name sounds good to me.

kbarros commented:
Are we good to merge now with some minor renamings? We can rename more later in a follow up.

Hao-Phys commented:
What are the renamings we need to do right now?

ddahlbom commented:
I'm just thinking about making the API as similar as possible to what it's final form will be and making it so that the FeI2 example will look as clean as possible. In this spirit, and based on the discussion above, I would suggest:

1) `lswt_unpolarized_INS_spec` -> `intensities`.
2) `lswt_dispersion_relation` -> `dispersion`.
3) `lswt_dynamical_spin_structure_factor!` -> `dssf`. It would also be good if the `Sαβ_matrix` could be allocated internally. (Another possible name for this function would be `dynamic_structure_factor`, but I think it's a bit odd that we would briefly have both `dynamic_structure_factor` and `DynamicStructureFactor`.)
4) `SpinWaveFields` -> `SpinWave`

The naming/interface for (3) and (4) will probably be subject to the most changes later on, but I think those changes would be good for now.

kbarros commented:
These suggestions look good to me. Thanks all, let's get it merged 👍

Kipton Barros, from phone

On Sun, Mar 19, 2023, 12:52 David A. Dahlbom ***@***.***>
wrote:

> I'm just thinking about making the API as similar as possible to what it's
> final form will be and making it so that the FeI2 example will look as
> clean as possible. In this spirit, and based on the discussion above, I
> would suggest:
>
>    1. lswt_unpolarized_INS_spec -> intensities.
>    2. lswt_dispersion_relation -> dispersion.
>    3. lswt_dynamical_spin_structure_factor! -> dssf. It would also be
>    good if the Sαβ_matrix could be allocated internally. (Another
>    possible name for this function would be dynamic_structure_factor, but
>    I think it's a bit odd that we would briefly have both
>    dynamic_structure_factor and DynamicStructureFactor.)
>    4. SpinWaveFields -> SpinWave
>
> The naming/interface for (3) and (4) will probably be subject to the most
> changes later on, but I think those changes would be good for now.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SunnySuite/Sunny.jl/pull/68#issuecomment-1475319227>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACG3HM325ZFEKNIGS5AHSDW442VLANCNFSM6AAAAAAV57FKLQ>
> .
> You are receiving this because you commented.Message ID: <SunnySuite/Sunny
> .***@***.***>
>

Hao-Phys commented:
I have changed the function names based on David's suggestions. Moreover, I have added supports for `biquad` in the `:dipole` mode (and a corresponding test item in test_lswt.jl). We are ready to merge.
